sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala

```diff
@@ -68,63 +68,95 @@ case class EnsureRequirements(
+        assert(requiredDist.isInstanceOf[ClusteredDistribution],
+          s"Expected ClusteredDistribution but found ${requiredDist.getClass.getSimpleName}")
+        i -> children(i).outputPartitioning.createShuffleSpec(
+          requiredDist.asInstanceOf[ClusteredDistribution])
+      }).toMap
 
-    if (childrenNumPartitions.size > 1) {
-      // Get the number of partitions which is explicitly required by the distributions.
-      val requiredNumPartitions = {
-        val numPartitionsSet = childrenIndexes.flatMap {
-          index => requiredChildDistributions(index).requiredNumPartitions
-        }.toSet
-        assert(numPartitionsSet.size <= 1,
-          s"$requiredChildDistributions have incompatible requirements of the number of partitions")
-        numPartitionsSet.headOption
-      }
+      // Find out the shuffle spec that gives better parallelism.
+      //
+      // NOTE: this is not optimal for the case when there are more than 2 children. Consider:
+      //   (10, 10, 11)
+      // it's better to pick 10 in this case since we only need to shuffle one side - we'd need to
+      // shuffle two sides if we pick 11.
+      //
+      // However this should be sufficient for now since in Spark nodes with multiple children
+      // always have exactly 2 children.
 
-      // If there are non-shuffle children that satisfy the required distribution, we have
-      // some tradeoffs when picking the expected number of shuffle partitions:
-      // 1. We should avoid shuffling these children.
-      // 2. We should have a reasonable parallelism.
-      val nonShuffleChildrenNumPartitions =
-        childrenIndexes.map(children).filterNot(_.isInstanceOf[ShuffleExchangeExec])
-          .map(_.outputPartitioning.numPartitions)
-      val expectedChildrenNumPartitions = if (nonShuffleChildrenNumPartitions.nonEmpty) {
-        if (nonShuffleChildrenNumPartitions.length == childrenIndexes.length) {
-          // Here we pick the max number of partitions among these non-shuffle children.
-          nonShuffleChildrenNumPartitions.max
+      // Whether we should consider `spark.sql.shuffle.partitions` and ensure enough parallelism
+      // during shuffle. To achieve a good trade-off between parallelism and shuffle cost, we only
+      // consider the minimum parallelism iff ALL children need to be re-shuffled.
+      //
+      // A child is considered to be re-shuffled iff:
+      //   1. It can't create partitioning by itself, i.e., `canCreatePartitioning` returns false.
+      //   2. It already has `ShuffleExchangeExec`.
+      //
+      // On the other hand, in scenarios such as:
+      //   HashPartitioning(5) <-> HashPartitioning(6)
+      // while `spark.sql.shuffle.partitions` is 10, we'll only re-shuffle the left side and make it
+      // HashPartitioning(6).
+      val canIgnoreMinPartitions = specs.exists(p =>
+        p._2.canCreatePartitioning && !children(p._1).isInstanceOf[ShuffleExchangeExec]
```

Comment:
Should we'd better use `ShuffleExchangeLike` or `ShuffleExchangeExec`? Will this work with customized columnar shuffle operator other than `ShuffleExchangeExec`?

Reply:
+1

Reply:
Hmm this is a bit tricky since later on we want to reuse a existing `ShuffleExchangeSpec` by just overwriting its number of partitions:

```scala
child match {
  case ShuffleExchangeExec(_, c, so) => ShuffleExchangeExec(newPartitioning, c, so)
  case _ => ShuffleExchangeExec(newPartitioning, child)
}
```

do we have any rule to remove the extra shuffle operators?

Reply:
> do we have any rule to remove the extra shuffle operators?

In Spark itself we don't, but the custom columnar rules (`ApplyColumnarRulesAndInsertTransitions`) after `EnsureRequirements` can do arbitrary logic, e.g., remove extra shuffle operators.